dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -690,138 +678,85 @@ impl<'a, 'tcx: 'a> Visitor<'tcx> for CollectFunctionInstrumentationPoints<'a, 't
+                let transfer_kind = if let &ty::FnDef(def_id, _) = func_kind {
+                    TransferKind::Arg(self.tcx.def_path_hash(def_id).convert())
+                } else {
+                    TransferKind::None
+                };
                 if !is_hook {
                     for arg in args {
                         if let Some(place) = arg.place() {
-                            if is_shared_or_unsafe_ptr(place.ty(self.body, self.tcx).ty) {
-                                println!(
-                                    "visiting terminator arg {:?}, {:?}, {:?}, {:?}",
-                                    place,
-                                    place.ty(self.body, self.tcx).ty,
-                                    InstrumentationOperand::from_type(
+                            let place_ty = place.ty(self.body, self.tcx).ty;
+                            if is_shared_or_unsafe_ptr(place_ty) {
+                                InstrumentationPoint::new(location, arg_fn)
+                                    .arg(InstrumentationArg::from_type(
                                         Operand::Copy(place),
-                                        &place.ty(self.body, self.tcx).ty,
-                                    ),
-                                    terminator
-                                );
-                                // Block terminator is a non-hook fn call; trace any raw-ptr args
-                                self.add_instrumentation_point(
-                                    location,
-                                    arg_fn,
-                                    vec![InstrumentationOperand::from_type(
-                                        Operand::Copy(place),
-                                        &place.ty(self.body, self.tcx).ty,
-                                    )],
-                                    false,
-                                    false,
-                                    EventMetadata {
-                                        source: Some(to_mir_place(&place)),
-                                        destination: Some(MirPlace {
-                                            local: arg_local,
-                                            projection: vec![],
-                                        }),
-                                        transfer_kind,
-                                    },
-                                );
+                                        &place_ty,
+                                    ))
+                                    .source(&place)
+                                    .dest(&callee_arg)
+                                    .transfer(transfer_kind)
+                                    .add(self);
                             }
                         }
-                        arg_local.index += 1;
+                        callee_arg = Local::from(callee_arg.local.as_u32() + 1).into();
```

Comment:
```suggestion
                        callee_arg.local.increment_by(1);
```

Requires `use rustc_index::vec::Idx` (separate suggestion).